File: app-js/src/main/scala/com/shocktrade/javascript/discover/DiscoverController.scala
package com.shocktrade.javascript.discover

import org.scalajs.angularjs._
import org.scalajs.angularjs.toaster.Toaster
import org.scalajs.angularjs.cookies.Cookies
import org.scalajs.angularjs.http.HttpResponse
import org.scalajs.nodejs.util.ScalaJsHelper._
import org.scalajs.angularjs.{angular, injected}
import com.shocktrade.javascript._
import com.shocktrade.javascript.dialogs.NewOrderDialogController.NewOrderDialogResult
import com.shocktrade.javascript.dialogs.{NewOrderDialog, NewOrderParams}
import com.shocktrade.javascript.discover.DiscoverController._
import com.shocktrade.javascript.models.{FullQuote, OrderQuote}
import com.shocktrade.javascript.profile.ProfileService
import org.scalajs.dom.console

import scala.concurrent.duration._
import scala.scalajs.concurrent.JSExecutionContext.Implicits.queue
import scala.scalajs.js
import scala.scalajs.js.JSConverters._
import scala.util.{Failure, Success}

/**
  * Discover Controller
  * @author lawrence.daniels@gmail.com
  */
class DiscoverController($scope: DiscoverControllerScope, $cookies: Cookies, $location: Location, $q: Q,
                         $routeParams: DiscoverRouteParams, $timeout: Timeout, toaster: Toaster,
                         @injected("MarketStatus") marketStatus: MarketStatusService,
                         @injected("MySessionService") mySession: MySessionService,
                         @injected("NewOrderDialog") newOrderDialog: NewOrderDialog,
                         @injected("ProfileService") profileService: ProfileService,
                         @injected("QuoteService") quoteService: QuoteService)
  extends AutoCompletionController($scope, $q, quoteService) with GlobalLoading {

  private var usMarketStatus: Either[MarketStatus, Boolean] = Right(false)

  // setup the public variables
  $scope.ticker = null
  $scope.q = FullQuote(active = true)

  // define the display options
  $scope.options = DiscoverOptions(range = $cookies.getOrElse("chart_range", "5d"))
  $scope.expanders = Expanders

  ///////////////////////////////////////////////////////////////////////////
  //          Public Function
  ///////////////////////////////////////////////////////////////////////////

  $scope.expandSection = (aModule: js.UndefOr[ModuleExpander]) => aModule foreach { module =>
    module.expanded = !module.expanded
  }

  $scope.popupNewOrderDialog = (aSymbol: js.UndefOr[String]) => aSymbol map { symbol =>
    newOrderDialog.popup(NewOrderParams(symbol = symbol))
  }

  $scope.loadQuote = (ticker: js.Dynamic) => {
    console.log(s"Loading symbol ${angular.toJson(ticker, pretty = false)}")

    // determine the symbol
    val symbol = (if (isDefined(ticker.symbol)) ticker.symbol.asOpt[String]
    else {
      ticker.asOpt[String] map { _ticker =>
        val index = _ticker.indexOf(" ")
        if (index == -1) _ticker else _ticker.substring(0, index)
      }
    }) map(_.toUpperCase)

    symbol foreach updateQuote
  }

  private def updateQuote(ticker: String) {
    // get the symbol (e.g. "AAPL - Apple Inc")
    val symbol = if (ticker.contains(" ")) ticker.substring(0, ticker.indexOf(" ")).trim else ticker

    // load the quote
    asyncLoading($scope)(quoteService.getStockQuote(symbol)) onComplete {
      case Success(quote) if isDefined(quote.symbol) =>
        // capture the quote
        $scope.q = quote
        $scope.ticker = s"${quote.symbol} - ${quote.name}"

        // update the address bar
        $location.search("symbol", quote.symbol)

        // store the last symbol
        $cookies.put(LastSymbolCookie, quote.symbol)

        // add the symbol to the Recently-viewed Symbols
        mySession.addRecentSymbol(symbol)

        // load the trading history
        $scope.tradingHistory = null
        if ($scope.expanders(6).expanded) {
          $scope.expandSection($scope.expanders(6))
        }

      case Success(quote) =>
        console.log(s"quote = ${angular.toJson(quote)}")
        toaster.warning(s"Symbol not found")

      case Failure(e) =>
        console.error(s"Failed to retrieve quote: ${e.getMessage}")
        toaster.error(s"Error loading quote $symbol")
    }
  }

  ///////////////////////////////////////////////////////////////////////////
  //          ETF Holdings / Products
  ///////////////////////////////////////////////////////////////////////////

  $scope.hasHoldings = (aQuote: js.UndefOr[FullQuote]) => aQuote exists { q =>
    q.legalType.exists(_ == "ETF") && q.products.exists(_.nonEmpty)
  }

  ///////////////////////////////////////////////////////////////////////////
  //          Symbols - Favorites
  ///////////////////////////////////////////////////////////////////////////

  $scope.addFavoriteSymbol = (aSymbol: js.UndefOr[String]) => {
    for {
      symbol <- aSymbol
      userId <- mySession.userProfile._id
    } yield profileService.addFavoriteSymbol(userId, symbol)
  }

  $scope.isFavorite = (aSymbol: js.UndefOr[String]) => aSymbol.exists(mySession.isFavoriteSymbol)

  $scope.removeFavoriteSymbol = (aSymbol: js.UndefOr[String]) => {
    for {
      symbol <- aSymbol
      userId <- mySession.userProfile._id
    } yield profileService.removeFavoriteSymbol(userId, symbol)
  }

  ///////////////////////////////////////////////////////////////////////////
  //          Symbols - Recently Viewed
  ///////////////////////////////////////////////////////////////////////////

  $scope.addRecentSymbol = (aSymbol: js.UndefOr[String]) => {
    for {
      symbol <- aSymbol
      userId <- mySession.userProfile._id
    } yield profileService.addRecentSymbol(userId, symbol)
  }

  $scope.isRecentSymbol = (symbol: js.UndefOr[String]) => {
    symbol.exists(mySession.isRecentSymbol)
  }

  $scope.removeRecentSymbol = (aSymbol: js.UndefOr[String]) => {
    for {
      symbol <- aSymbol
      userId <- mySession.userProfile._id
    } yield profileService.removeRecentSymbol(userId, symbol)
  }

  ///////////////////////////////////////////////////////////////////////////
  //          Risk Functions
  ///////////////////////////////////////////////////////////////////////////

  $scope.getBetaClass = (beta: js.UndefOr[Double]) => beta map {
    case b if b > 1.3 || b < -1.3 => "volatile_red"
    case b if b >= 0.0 => "volatile_green"
    case b if b < 0 => "volatile_yellow"
    case _ => js.undefined
  }

  $scope.getRiskClass = (riskLevel: js.UndefOr[String]) => riskLevel map {
    case rs if isDefined(rs) && rs.nonEmpty => s"risk_${rs.toLowerCase}"
    case _ => js.undefined
  }

  $scope.getRiskDescription = (riskLevel: js.UndefOr[String]) => riskLevel map {
    case "Low" => "Generally recommended for investment"
    case "Medium" => "Not recommended for inexperienced investors"
    case "High" => "Not recommended for investment"
    case "Unknown" => "The risk level could not be determined"
    case _ => "The risk level could not be determined"
  }

  ///////////////////////////////////////////////////////////////////////////
  //          Market Status Functions
  ///////////////////////////////////////////////////////////////////////////

  $scope.isUSMarketsOpen = () => {
    usMarketStatus match {
      case Left(status) => Option(status.active).orUndefined
      case Right(loading) =>
        if (!loading) {
          usMarketStatus = Right(true)
          console.log("Retrieving market status...")
          marketStatus.getMarketStatus onComplete {
            case Success(status) =>
              // {"stateChanged":false,"active":false,"sysTime":1392092448795,"delay":-49848795,"start":1392042600000,"end":1392066000000}
              // retrieve the delay in milliseconds from the server
              var delay = status.delay
              if (delay < 0) {
                delay = Math.max(status.end - status.sysTime, 300000)
              }

              // set the market status
              console.log(s"US Markets are ${if (status.active) "Open" else "Closed"}; Waiting for $delay msec until next trading start...")
              usMarketStatus = Left(status)

              // update the status after delay
              console.log(s"Re-loading market status in ${status.delay.minutes}")
              $timeout(() => usMarketStatus = Right(false), 5.minutes)

            case Failure(e) =>
              toaster.error("Failed to retrieve market status")
              console.error(s"Failed to retrieve market status: ${e.getMessage}")
          }
        }
        js.undefined
    }
  }

  ///////////////////////////////////////////////////////////////////////////
  //          Initialization
  ///////////////////////////////////////////////////////////////////////////

  // load the symbol
  if ($scope.q.symbol.isEmpty) {
    // get the symbol
    val symbol = $routeParams.symbol getOrElse $cookies.getOrElse(LastSymbolCookie, mySession.getMostRecentSymbol)

    // load the symbol
    updateQuote(symbol)
  }

  ///////////////////////////////////////////////////////////////////////////
  //          Event Listeners
  ///////////////////////////////////////////////////////////////////////////

  // setup the chart range
  $scope.$watch("options.range", (newValue: js.UndefOr[Any], oldValue: js.Any) => newValue.foreach($cookies.put("chart_range", _)))

}

/**
  * Discover Controller
  */
object DiscoverController {
  val LastSymbolCookie = "QuoteService_lastSymbol"

  def isPerformanceRisk: js.Function1[js.UndefOr[FullQuote], Boolean] = (aQuote: js.UndefOr[FullQuote]) => aQuote.exists { q =>
    q.high52Week.nonEmpty || q.low52Week.nonEmpty || q.change52Week.nonEmpty ||
      q.movingAverage50Day.nonEmpty || q.movingAverage200Day.nonEmpty ||
      q.change52WeekSNP500.nonEmpty || q.beta.nonEmpty
  }

  def isIncomeStatement: js.Function1[js.UndefOr[FullQuote], Boolean] = (aQuote: js.UndefOr[FullQuote]) => aQuote.exists { q =>
    q.revenue.nonEmpty || q.revenuePerShare.nonEmpty || q.revenueGrowthQuarterly.nonEmpty ||
      q.grossProfit.nonEmpty || q.EBITDA.nonEmpty || q.netIncomeAvailToCommon.nonEmpty ||
      q.dilutedEPS.nonEmpty || q.earningsGrowthQuarterly.nonEmpty
  }

  def isBalanceSheet: js.Function1[js.UndefOr[FullQuote], Boolean] = (aQuote: js.UndefOr[FullQuote]) => aQuote.exists { q =>
    q.totalCash.nonEmpty || q.totalDebt.nonEmpty || q.currentRatio.nonEmpty ||
      q.totalCashPerShare.nonEmpty || q.totalDebtOverEquity.nonEmpty || q.bookValuePerShare.nonEmpty ||
      q.returnOnAssets.nonEmpty || q.profitMargin.nonEmpty || q.mostRecentQuarterDate.nonEmpty ||
      q.returnOnEquity.nonEmpty || q.operatingMargin.nonEmpty || q.fiscalYearEndDate.nonEmpty
  }

  def isValuationMeasures: js.Function1[js.UndefOr[FullQuote], Boolean] = (aQuote: js.UndefOr[FullQuote]) => aQuote.exists { q =>
    q.enterpriseValue.nonEmpty || q.trailingPE.nonEmpty || q.forwardPE.nonEmpty ||
      q.pegRatio.nonEmpty || q.priceOverSales.nonEmpty || q.priceOverBookValue.nonEmpty ||
      q.enterpriseValueOverRevenue.nonEmpty || q.enterpriseValueOverEBITDA.nonEmpty ||
      q.operatingCashFlow.nonEmpty || q.leveredFreeCashFlow.nonEmpty
  }

  def isShareStatistics: js.Function1[js.UndefOr[FullQuote], Boolean] = (aQuote: js.UndefOr[FullQuote]) => aQuote.exists { q =>
    q.avgVolume3Month.nonEmpty || q.avgVolume10Day.nonEmpty || q.sharesOutstanding.nonEmpty ||
      q.sharesFloat.nonEmpty || q.pctHeldByInsiders.nonEmpty || q.pctHeldByInstitutions.nonEmpty ||
      q.sharesShort.nonEmpty || q.shortRatio.nonEmpty || q.shortPctOfFloat.nonEmpty || q.sharesShortPriorMonth.nonEmpty
  }

  def isDividendsSplits: js.Function1[js.UndefOr[FullQuote], Boolean] = (aQuote: js.UndefOr[FullQuote]) => aQuote.exists { q =>
    q.forwardAnnualDividendRate.nonEmpty || q.forwardAnnualDividendYield.nonEmpty ||
      q.trailingAnnualDividendYield.nonEmpty || q.divYield5YearAvg.nonEmpty ||
      q.payoutRatio.nonEmpty || q.dividendDate.nonEmpty || q.exDividendDate.nonEmpty ||
      q.lastSplitFactor.nonEmpty || q.lastSplitDate.nonEmpty
  }

  // define the Quote module expanders
  val Expanders = js.Array(
    ModuleExpander(title = "Performance & Risk",
      url = "/assets/views/discover/expanders/price_performance.htm",
      icon = "fa-line-chart",
      expanded = false,
      visible = isPerformanceRisk),
    ModuleExpander(title = "Income Statement",
      url = "/assets/views/discover/expanders/income_statement.htm",
      icon = "fa-money",
      expanded = false,
      visible = isIncomeStatement),
    ModuleExpander(title = "Balance Sheet",
      url = "/assets/views/discover/expanders/balance_sheet.htm",
      icon = "fa-calculator",
      expanded = false,
      visible = isBalanceSheet),
    ModuleExpander(title = "Valuation Measures",
      url = "/assets/views/discover/expanders/valuation_measures.htm",
      icon = "fa-gears",
      expanded = false,
      visible = isValuationMeasures),
    ModuleExpander(title = "Share Statistics",
      url = "/assets/views/discover/expanders/share_statistics.htm",
      icon = "fa-bar-chart",
      expanded = false,
      visible = isShareStatistics),
    ModuleExpander(title = "Dividends & Splits",
      url = "/assets/views/discover/expanders/dividends_splits.htm",
      icon = "fa-cut",
      expanded = false,
      visible = isDividendsSplits),
    ModuleExpander(title = "Historical Quotes",
      url = "/assets/views/discover/quotes/trading_history.htm",
      icon = "fa-calendar",
      expanded = false))

}

/**
  * Discover Controller Scope
  */
@js.native
trait DiscoverControllerScope extends AutoCompletionControllerScope {
  // variables
  var expanders: js.Array[ModuleExpander]
  var options: DiscoverOptions
  var q: OrderQuote
  var ticker: String
  var tradingHistory: js.Array[js.Object]

  // functions
  var expandSection: js.Function1[js.UndefOr[ModuleExpander], Unit]
  var hasHoldings: js.Function1[js.UndefOr[FullQuote], Boolean]
  var isUSMarketsOpen: js.Function0[js.UndefOr[Boolean]]
  var loadQuote: js.Function1[js.Dynamic, Unit]
  var popupNewOrderDialog: js.Function1[js.UndefOr[String], js.UndefOr[js.Promise[NewOrderDialogResult]]]

  // favorite quote functions
  var addFavoriteSymbol: js.Function1[js.UndefOr[String], js.UndefOr[HttpResponse[js.Dynamic]]]
  var isFavorite: js.Function1[js.UndefOr[String], Boolean]
  var removeFavoriteSymbol: js.Function1[js.UndefOr[String], js.UndefOr[HttpResponse[js.Dynamic]]]

  // recently-viewed quote functions
  var addRecentSymbol: js.Function1[js.UndefOr[String], js.UndefOr[HttpResponse[js.Dynamic]]]
  var isRecentSymbol: js.Function1[js.UndefOr[String], Boolean]
  var removeRecentSymbol: js.Function1[js.UndefOr[String], js.UndefOr[HttpResponse[js.Dynamic]]]

  // risk functions
  var getBetaClass: js.Function1[js.UndefOr[Double], js.UndefOr[Object]]
  var getRiskClass: js.Function1[js.UndefOr[String], js.UndefOr[Object]]
  var getRiskDescription: js.Function1[js.UndefOr[String], js.UndefOr[String]]

}

/**
  * Discover Options
  */
@js.native
trait DiscoverOptions extends js.Object {
  var range: String
}

/**
  * Discover Options Singleton
  */
object DiscoverOptions {

  def apply(range: String) = {
    val options = New[DiscoverOptions]
    options.range = range
    options
  }

}

/**
  * Discover Route Parameters
  */
@js.native
trait DiscoverRouteParams extends js.Object {
  var symbol: js.UndefOr[String]

}

File: app-play/app/assets/javascripts/discover/DiscoverController.js
(function () {
    var app = angular.module('shocktrade');

    /**
     * Discover Controller
     * @author lawrence.daniels@gmail.com
     */
    app.controller('DiscoverController', ['$scope', '$cookieStore', '$interval', '$log', '$routeParams', '$timeout', 'toaster', 'FavoriteSymbols', 'HeldSecurities', 'MarketStatus', 'MySession', 'NewOrderDialog', 'QuoteService', 'RecentSymbols',
        function ($scope, $cookieStore, $interval, $log, $routeParams, $timeout, toaster, FavoriteSymbols, HeldSecurities, MarketStatus, MySession, NewOrderDialog, QuoteService, RecentSymbols) {

            // setup a private loading variable
            $scope.loading = false;

            // setup the public variables
            $scope.marketClock = (new Date()).toTimeString();
            $scope.ticker = null;
            $scope.q = {active: true};

            // define the display options
            $scope.options = {
                range: $cookieStore.get("chart_range") || "5d"
            };

            // setup the chart range
            $scope.$watch("options.range", function (newValue, oldValue) {
                $cookieStore.put("chart_range", newValue);
            });

            // setup filtered quotes & trading history
            $scope.filterQuotes = [];
            $scope.tradingHistory = null;
            $scope.selectedTradingHistory = null;

            // define the Quote module expanders
            $scope.expanders = [{
                title: "Performance & Risk",
                url: "/assets/views/discover/quotes/expanders/price_performance.htm",
                icon: "fa-line-chart",
                expanded: false,
                visible: function (q) {
                    return q.high52Week || q.low52Week || q.change52Week || q.movingAverage50Day ||
                        q.movingAverage200Day || q.change52WeekSNP500 || q.beta;
                }
            }, {
                title: "Income Statement",
                url: "/assets/views/discover/quotes/expanders/income_statement.htm",
                icon: "fa-money",
                expanded: false,
                visible: function (q) {
                    return q.revenue || q.revenuePerShare || q.revenueGrowthQuarterly || q.grossProfit ||
                        q.EBITDA || q.netIncomeAvailToCommon || q.dilutedEPS || q.earningsGrowthQuarterly;
                }
            }, {
                title: "Balance Sheet",
                url: "/assets/views/discover/quotes/expanders/balanace_sheet.htm",
                icon: "fa-calculator",
                expanded: false,
                visible: function (q) {
                    return q.totalCash || q.totalDebt || q.currentRatio || q.totalCashPerShare ||
                        q.totalDebtOverEquity || q.bookValuePerShare || q.returnOnAssets ||
                        q.profitMargin || q.mostRecentQuarterDate || q.returnOnEquity ||
                        q.operatingMargin || q.fiscalYearEndDate;
                }
            }, {
                title: "Valuation Measures",
                url: "/assets/views/discover/quotes/expanders/valuation_measures.htm",
                icon: "fa-gears",
                expanded: false,
                visible: function (q) {
                    return q.enterpriseValue || q.trailingPE || q.forwardPE || q.pegRatio || q.priceOverSales ||
                        q.priceOverBookValue || q.enterpriseValueOverRevenue || q.enterpriseValueOverEBITDA ||
                        q.operatingCashFlow || q.leveredFreeCashFlow;
                }
            }, {
                title: "Share Statistics",
                url: "/assets/views/discover/quotes/expanders/share_statistics.htm",
                icon: "fa-bar-chart",
                expanded: false,
                visible: function (q) {
                    return q.avgVolume3Month || q.avgVolume10Day || q.sharesOutstanding || q.sharesFloat ||
                        q.pctHeldByInsiders || q.pctHeldByInstitutions || q.sharesShort || q.shortRatio ||
                        q.shortPctOfFloat || q.sharesShortPriorMonth;
                }
            }, {
                title: "Dividends & Splits",
                url: "/assets/views/discover/quotes/expanders/dividends_splits.htm",
                icon: "fa-cut",
                expanded: false,
                visible: function (q) {
                    return q.forwardAnnualDividendRate || q.forwardAnnualDividendYield ||
                        q.trailingAnnualDividendYield || q.divYield5YearAvg || q.payoutRatio ||
                        q.dividendDate || q.exDividendDate || q.lastSplitFactor || q.lastSplitDate;
                }
            }, {
                title: "Historical Quotes",
                url: "/assets/views/discover/quotes/trading_history.htm",
                icon: "fa-calendar",
                expanded: false,
                loading: false,
                onExpand: function (callback) {
                    if ($scope.tradingHistory === null && $scope.q.assetType === 'Common Stock') {
                        $scope.loadTradingHistory($scope.q.symbol, callback);
                    }
                }
            }];

            $scope.autoCompleteSymbols = function (searchTerm) {
                return QuoteService.autoCompleteSymbols(searchTerm, 20)
                    .success(function (response) {
                        return response;
                    })
                    .error(function (err) {
                        toaster.pop("error", "Error auto-completing symbol", null)
                    });
            };

            /**
             * Initializes the module
             */
            $scope.init = function () {
                // setup market status w/updates
                $interval(function () {
                    $scope.marketClock = (new Date()).toTimeString();
                }, 1000);

                // setup the market status updates
                setupMarketStatusUpdates();
            };

            $scope.popupNewOrderDialog = function (symbol) {
                NewOrderDialog.popup({symbol: symbol});
            };

            $scope.expandSection = function (module) {
                module.expanded = !module.expanded;
                if (module.expanded && module.onExpand) {
                    module.loading = true;
                    var promise = $timeout(function () {
                        module.loading = false;
                    }, 3000);
                    module.onExpand(function () {
                        module.loading = false;
                        $timeout.cancel(promise);
                    });
                }
            };

            $scope.addFavoriteSymbol = function (symbol) {
                FavoriteSymbols.add(symbol);
            };

            $scope.isFavorite = function (symbol) {
                return FavoriteSymbols.isFavorite(symbol);
            };

            $scope.removeFavoriteSymbol = function (symbol) {
                FavoriteSymbols.remove(symbol);
            };

            $scope.hasHoldings = function (q) {
                return q && q.products && (q.legalType === 'ETF') && (q.products.length !== 0);
            };

            $scope.removeRecentSymbol = function (symbol) {
                RecentSymbols.remove(symbol);
            };

            $scope.rowClass = function (column, row) {
                return ( column === "symbol" ) ? row['exchange'] : column;
            };

            $scope.columnAlign = function (column) {
                return ( column === "symbol" ) ? "left" : "right";
            };

            $scope.isOpen = function (filter) {
                return (filter.name === "Favorites");
            };

            $scope.getMatchedAssetIcon = function (q) {
                //console.log("q = " + angular.toJson(q));
                return "/assets/images/asset_types/stock.png";
            };

            $scope.getRiskClass = function (riskLevel) {
                return riskLevel ? "risk_" + riskLevel.toLowerCase() : null;
            };

            $scope.getRiskDescription = function (riskLevel) {
                if (riskLevel === "Low") return "Generally recommended for investment";
                else if (riskLevel === "Medium") return "Not recommended for inexperienced investors";
                else if (riskLevel === "High") return "Not recommended for investment";
                else if (riskLevel === "Unknown") return "The risk level could not be determined";
                else return "The risk level could not be determined";
            };

            $scope.getBetaClass = function (beta) {
                if (beta == null) return "";
                else if (beta > 1.3 || beta < -1.3) return "volatile_red";
                else if (beta >= 0.0) return "volatile_green";
                else if (beta < 0) return "volatile_yellow";
                else return "";
            };

            $scope.loadTickerQuote = function (_ticker) {
                var ticker = $("#stockTicker").val() || _ticker;
                $scope.loadQuote(ticker);
            };

            $scope.loadQuote = function (ticker) {
                $log.info("Loading symbol " + angular.toJson(ticker));

                // setup the loading animation
                $scope.startLoading();

                // determine the symbol
                var symbol = null;
                if (ticker.symbol) {
                    symbol = ticker.symbol.toUpperCase();
                }
                else {
                    var index = ticker.indexOf(' ');
                    symbol = (index == -1 ? ticker : ticker.substring(0, index)).toUpperCase();
                }

                // load the quote
                QuoteService.getStockQuote(symbol)
                    .success(function (quote) {
                        if (quote) {
                            // capture the quote
                            $scope.q = quote;
                            $scope.ticker = quote.symbol + " - " + quote.name;

                            // store the last symbol
                            $cookieStore.put("QuoteService_lastSymbol", quote.symbol);

                            // add the symbol to the Recently-viewed Symbols
                            RecentSymbols.add(symbol);

                            // get the risk level
                            QuoteService.getRiskLevel(symbol)
                                .success(function (response) {
                                    quote.riskLevel = response;
                                })
                                .error(function (response) {
                                    toaster.pop('error', 'Error!', "Error retrieving risk level for " + symbol);
                                });

                            // load the trading history
                            $scope.tradingHistory = null;
                            if ($scope.expanders[6].expanded) {
                                $scope.expandSection($scope.expanders[6]);
                            }
                        }
                        else {
                            toaster.pop('error', 'Error!', "No quote found for " + symbol);
                            console.log("Empty quote? " + angular.toJson(quote));
                        }

                        // disabling the loading status
                        $scope.stopLoading();
                    })
                    .error(function (response) {
                        $log.error("Failed to retrieve quote: " + response.status);
                        $scope.stopLoading();
                        toaster.pop('error', 'Error!', "Error loading quote " + symbol);
                    });
            };

            $scope.loadTradingHistory = function (symbol, callback) {
                QuoteService.getTradingHistory(symbol)
                    .success(function (results) {
                        $scope.tradingHistory = results;
                        if (callback) callback();
                    })
                    .error(function (response) {
                        toaster.pop('error', 'Error!', "Error loading trading history for " + symbol);
                        if (callback) callback();
                    });
            };

            $scope.loadFilterQuotes = function (filter, index) {
                $scope.loading = true;
                QuoteService.getFilterQuotes(filter)
                    .success(function (quotes) {
                        $scope.loading = false;
                        filter.rows = quotes;
                    })
                    .error(function (err) {
                        $scope.loading = false;
                        toaster.pop('error', 'Error!', "Error loading filter " + filter.name);
                    });
            };

            $scope.tradingActive = function (time) {
                return (new Date()).getTime();
            };

            $scope.selectTradingHistory = function (t) {
                $scope.selectedTradingHistory = t;
            };

            $scope.hasSelectedTradingHistory = function () {
                return $scope.selectedTradingHistory != null;
            };

            $scope.isSelectedTradingHistory = function (t) {
                return $scope.selectedTradingHistory === t;
            };

            function setupMarketStatusUpdates() {
                $scope.usMarketsOpen = null;
                $log.info("Retrieving market status...");
                MarketStatus.getMarketStatus()
                    .success(function (status) {
                        // retrieve the delay in milliseconds from the server
                        var delay = status.delay;
                        if (delay < 0) {
                            delay = status.end - status.sysTime;
                            if (delay <= 300000) {
                                delay = 300000; // 5 minutes
                            }
                        }

                        // set the market status
                        $log.info("US Markets are " + (status.active ? 'Open' : 'Closed') + "; Waiting for " + delay + " msec until next trading start...");
                        setTimeout(function () {
                            $scope.usMarketsOpen = status.active;
                        }, 750);

                        // wait for the delay, then call recursively
                        setTimeout(function () {
                            setupMarketStatusUpdates();
                        }, delay);
                    })
                    .error(function(err) {
                        toaster.pop('error', 'Failed to retrieve market status');
                    });
            }

            ///////////////////////////////////////////////////////////////////////////
            //          Initialization
            ///////////////////////////////////////////////////////////////////////////

            // load the symbol
            (function () {
                if (!$scope.q.symbol) {
                    // get the symbol
                    var symbol = $routeParams.symbol || $cookieStore.get('QuoteService_lastSymbol') || RecentSymbols.getLast();

                    // load the symbol
                    $scope.loadQuote(symbol);
                }
            })();

        }]);

})();